libs/wire-api/src/Wire/API/User.hs

```diff
@@ -1356,6 +1358,31 @@ instance FromJSON DeletionCodeTimeout where
   parseJSON = A.withObject "DeletionCodeTimeout" $ \o ->
     DeletionCodeTimeout <$> o A..: "expires_in"
 
+data DeleteUserResult = NoUser | AccountAlreadyDeleted | AccountDeleted
+  deriving (Eq, Show)
+
+instance ToJSON DeleteUserResult where
+  toJSON t = A.object ["tag" A..= toTag t]
+    where
+      toTag :: DeleteUserResult -> A.Value
+      toTag NoUser = "no-user"
+      toTag AccountAlreadyDeleted = "already-deleted"
+      toTag AccountDeleted = "deleted"
+
+instance FromJSON DeleteUserResult where
+  parseJSON (A.Object o) = do
+    tagString <- o A..: "tag"
+    case fromTag tagString of
+      Just t -> pure t
+      Nothing -> A.parseFail $ "Unknown tag: " ++ tagString
```

Comment:
```suggestion
    maybe (A.parseFail $ "Unknown tag: " ++ tagString) pure (fromTag tagString)
```

I'm not going to tell you which version I like better, just wanted to see for myself.  :)

Reply:
Well, I think your solution is more abstract and elegant, while mine is a bit easier to grasp (Go and ask some Java/Scala/Kotlin developers... :wink: ).

As the aeson instances aren't needed anymore (endpoint results are now only HTTP status codes), this debate will likely never be answered... :smile_cat: